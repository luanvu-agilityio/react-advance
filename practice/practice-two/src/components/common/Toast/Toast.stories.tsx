import type { Meta, StoryObj } from '@storybook/react'
import { ToastContainer, type Toast } from './Toast'
import { useEffect, useState } from 'react'
import { Button } from '@radix-ui/themes'

const meta: Meta<typeof ToastContainer> = {
  title: 'Components/Toast',
  component: ToastContainer,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof ToastContainer>

// Helper component to control toasts in Storybook
const ToastManager = ({
  initialToasts,
  showControls = false,
}: {
  initialToasts: Toast[]
  showControls?: boolean
}) => {
  const [toasts, setToasts] = useState<Toast[]>(initialToasts)

  const removeToast = (id: string) => {
    setToasts((prev) => prev.filter((toast) => toast.id !== id))
  }

  const addToast = (variant: Toast['variant']) => {
    const id = `toast-${Date.now()}`
    const newToast: Toast = {
      id,
      title: `${variant.charAt(0).toUpperCase() + variant.slice(1)} Toast`,
      description: `This is a ${variant} toast message`,
      variant,
    }
    setToasts((prev) => [...prev, newToast])
  }

  return (
    <div>
      {showControls && (
        <div style={{ display: 'flex', gap: '8px', marginBottom: '20px' }}>
          <Button onClick={() => addToast('success')}>Add Success</Button>
          <Button onClick={() => addToast('error')}>Add Error</Button>
          <Button onClick={() => addToast('warning')}>Add Warning</Button>
          <Button onClick={() => addToast('info')}>Add Info</Button>
        </div>
      )}

      <ToastContainer toasts={toasts} onRemoveToast={removeToast} />
    </div>
  )
}

// Single toast stories
export const SuccessToast: Story = {
  render: () => {
    const successToast: Toast[] = [
      {
        id: 'success-1',
        title: 'Success Toast',
        description: 'Your action was completed successfully!',
        variant: 'success',
      },
    ]
    return <ToastManager initialToasts={successToast} />
  },
}

export const ErrorToast: Story = {
  render: () => {
    const errorToast: Toast[] = [
      {
        id: 'error-1',
        title: 'Error Toast',
        description: 'Something went wrong!',
        variant: 'error',
      },
    ]
    return <ToastManager initialToasts={errorToast} />
  },
}

export const WarningToast: Story = {
  render: () => {
    const warningToast: Toast[] = [
      {
        id: 'warning-1',
        title: 'Warning Toast',
        description: 'Please review your information before proceeding.',
        variant: 'warning',
      },
    ]
    return <ToastManager initialToasts={warningToast} />
  },
}

export const InfoToast: Story = {
  render: () => {
    const infoToast: Toast[] = [
      {
        id: 'info-1',
        title: 'Info Toast',
        description: 'Just letting you know!',
        variant: 'info',
      },
    ]
    return <ToastManager initialToasts={infoToast} />
  },
}

// Toast without description
export const ToastWithoutDescription: Story = {
  render: () => {
    const toast: Toast[] = [
      {
        id: 'no-desc-1',
        title: 'Toast Without Description',
        variant: 'info',
      },
    ]
    return <ToastManager initialToasts={toast} />
  },
}

// Custom duration toast
export const CustomDurationToast: Story = {
  render: () => {
    const toast: Toast[] = [
      {
        id: 'long-duration-1',
        title: 'Long Duration Toast (10s)',
        description: 'This toast will stay for 10 seconds',
        variant: 'success',
        duration: 10000,
      },
    ]
    return <ToastManager initialToasts={toast} />
  },
}

// Multiple toasts
export const MultipleToasts: Story = {
  render: () => {
    const allVariantToasts: Toast[] = [
      {
        id: 'multi-success',
        title: 'Success Toast',
        description: 'Your action was completed successfully!',
        variant: 'success',
      },
      {
        id: 'multi-error',
        title: 'Error Toast',
        description: 'Something went wrong!',
        variant: 'error',
      },
      {
        id: 'multi-warning',
        title: 'Warning Toast',
        description: 'Please review your information before proceeding.',
        variant: 'warning',
      },
      {
        id: 'multi-info',
        title: 'Info Toast',
        description: 'Just letting you know!',
        variant: 'info',
      },
    ]
    return <ToastManager initialToasts={allVariantToasts} />
  },
}

// Interactive demo with controls
export const InteractiveDemo: Story = {
  render: () => {
    return <ToastManager initialToasts={[]} showControls={true} />
  },
}

// Auto-add toast when story loads
export const AutoShowToast: Story = {
  render: () => {
    // Create a component to handle the auto-showing toast
    const AutoShowToastComponent = () => {
      const [toasts, setToasts] = useState<Toast[]>([])

      useEffect(() => {
        // Add a toast when the component mounts
        const id = 'auto-toast'
        setToasts([
          {
            id,
            title: 'Auto-Shown Toast',
            description: 'This toast appears automatically',
            variant: 'success',
            duration: 5000,
          },
        ])
      }, [])

      const removeToast = (id: string) => {
        setToasts((prev) => prev.filter((toast) => toast.id !== id))
      }

      return <ToastContainer toasts={toasts} onRemoveToast={removeToast} />
    }

    return <AutoShowToastComponent />
  },
}
