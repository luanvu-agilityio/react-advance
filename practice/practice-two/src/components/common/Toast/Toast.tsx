import * as RadixToast from '@radix-ui/react-toast'

export type ToastVariant = 'success' | 'error' | 'warning' | 'info'

export interface Toast {
  id: string
  title: string
  description?: string
  variant: ToastVariant
  duration?: number
}

interface ToastContainerProps {
  toasts: Toast[]
  onRemoveToast: (id: string) => void
}

export const ToastContainer = ({
  toasts,
  onRemoveToast,
}: ToastContainerProps) => {
  return (
    <RadixToast.Provider>
      {toasts.map((toast) => (
        <ToastItem
          key={toast.id}
          toast={toast}
          onClose={() => onRemoveToast(toast.id)}
        />
      ))}
      <RadixToast.Viewport className="toast-viewport" />
    </RadixToast.Provider>
  )
}

interface ToastItemProps {
  toast: Toast
  onClose: () => void
}

const ToastItem = ({ toast, onClose }: ToastItemProps) => {
  const getColor = () => {
    switch (toast.variant) {
      case 'success':
        return '#2B9348'
      case 'error':
        return '#D00000'
      case 'warning':
        return '#E85D04'
      case 'info':
        return '#1E88E5'
      default:
        return '#1E88E5'
    }
  }

  return (
    <RadixToast.Root
      className={`toast-root toast-${toast.variant}`}
      duration={toast.duration ?? 3000}
      onOpenChange={(open) => {
        if (!open) onClose()
      }}
    >
      <div>
        <RadixToast.Title className="toast-title" style={{ color: getColor() }}>
          {toast.title}
        </RadixToast.Title>
        {toast.description && (
          <RadixToast.Description className="toast-description">
            {toast.description}
          </RadixToast.Description>
        )}
      </div>
      <RadixToast.Close className="toast-close">Ã—</RadixToast.Close>
    </RadixToast.Root>
  )
}
