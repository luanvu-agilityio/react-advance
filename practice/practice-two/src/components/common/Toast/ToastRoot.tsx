import { useToastStore } from '@stores/toastStore'
import { ToastContainer } from './Toast'

export const ToastRoot = () => {
  const toasts = useToastStore((state) => state.toasts)
  const removeToast = useToastStore((state) => state.removeToast)

  return <ToastContainer toasts={toasts} onRemoveToast={removeToast} />
}
