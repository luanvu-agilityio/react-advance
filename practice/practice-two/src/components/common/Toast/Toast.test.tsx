import { act, fireEvent, render, screen } from '@testing-library/react'
import { ToastContainer } from './Toast'
import type { Toast } from './Toast'

// SNAPSHOT TESTS FOR RENDERING

describe('Toast Component - Rendering', () => {
  // Sample toast data for testing
  const successToast: Toast = {
    id: 'test-id-1',
    title: 'Success Toast',
    description: 'This is a success message',
    variant: 'success',
  }

  const errorToast: Toast = {
    id: 'test-id-2',
    title: 'Error Toast',
    description: 'This is an error message',
    variant: 'error',
  }

  const warningToast: Toast = {
    id: 'test-id-3',
    title: 'Warning Toast',
    variant: 'warning',
  }

  const infoToast: Toast = {
    id: 'test-id-4',
    title: 'Info Toast',
    variant: 'info',
  }

  test('matches snapshot for success toast', () => {
    const { container } = render(
      <ToastContainer toasts={[successToast]} onRemoveToast={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
  })

  test('matches snapshot for error toast', () => {
    const { container } = render(
      <ToastContainer toasts={[errorToast]} onRemoveToast={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
  })

  test('matches snapshot for warning toast', () => {
    const { container } = render(
      <ToastContainer toasts={[warningToast]} onRemoveToast={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
  })

  test('matches snapshot for info toast', () => {
    const { container } = render(
      <ToastContainer toasts={[infoToast]} onRemoveToast={jest.fn()} />
    )
    expect(container).toMatchSnapshot()
  })

  test('renders toast with correct title', () => {
    render(<ToastContainer toasts={[successToast]} onRemoveToast={jest.fn()} />)
    expect(screen.getByText('Success Toast')).toBeInTheDocument()
  })

  test('renders toast with description when provided', () => {
    render(<ToastContainer toasts={[successToast]} onRemoveToast={jest.fn()} />)
    expect(screen.getByText('This is a success message')).toBeInTheDocument()
  })

  test('renders toast without description', () => {
    render(<ToastContainer toasts={[warningToast]} onRemoveToast={jest.fn()} />)
    expect(screen.getByText('Warning Toast')).toBeInTheDocument()
    expect(screen.queryByText(/this is a/i)).not.toBeInTheDocument()
  })

  test('renders multiple toasts when provided', () => {
    render(
      <ToastContainer
        toasts={[successToast, errorToast, warningToast]}
        onRemoveToast={jest.fn()}
      />
    )
    expect(screen.getByText('Success Toast')).toBeInTheDocument()
    expect(screen.getByText('Error Toast')).toBeInTheDocument()
    expect(screen.getByText('Warning Toast')).toBeInTheDocument()
  })

  test('applies correct color to toast title based on variant', () => {
    const { rerender } = render(
      <ToastContainer toasts={[successToast]} onRemoveToast={jest.fn()} />
    )

    // Check success color
    const successTitle = screen.getByText('Success Toast')
    expect(successTitle).toHaveStyle({ color: '#2B9348' })

    // Check error color
    rerender(<ToastContainer toasts={[errorToast]} onRemoveToast={jest.fn()} />)
    const errorTitle = screen.getByText('Error Toast')
    expect(errorTitle).toHaveStyle({ color: '#D00000' })

    // Check warning color
    rerender(
      <ToastContainer toasts={[warningToast]} onRemoveToast={jest.fn()} />
    )
    const warningTitle = screen.getByText('Warning Toast')
    expect(warningTitle).toHaveStyle({ color: '#E85D04' })

    // Check info color
    rerender(<ToastContainer toasts={[infoToast]} onRemoveToast={jest.fn()} />)
    const infoTitle = screen.getByText('Info Toast')
    expect(infoTitle).toHaveStyle({ color: '#1E88E5' })
  })
})

jest.useFakeTimers()

// INTERACTIVE BEHAVIOR TESTS
describe('Toast Component - Interactive', () => {
  afterEach(() => {
    jest.clearAllMocks()
    jest.clearAllTimers()
  })

  test('calls onRemoveToast when close button is clicked', async () => {
    const mockOnRemoveToast = jest.fn()
    const mockToast: Toast = {
      id: 'test-id-1',
      title: 'Test Toast',
      variant: 'success',
    }

    render(
      <ToastContainer toasts={[mockToast]} onRemoveToast={mockOnRemoveToast} />
    )

    // Use fireEvent instead of userEvent for simpler synchronous behavior
    const closeButton = screen.getByText('Ã—')
    fireEvent.click(closeButton)

    // Assert immediately after click without waiting
    expect(mockOnRemoveToast).toHaveBeenCalledWith('test-id-1')
  })

  test('automatically removes toast after custom duration', () => {
    const mockOnRemoveToast = jest.fn()
    const mockToast: Toast = {
      id: 'test-id-1',
      title: 'Test Toast',
      variant: 'success',
      duration: 1000, // 1 second
    }

    render(
      <ToastContainer toasts={[mockToast]} onRemoveToast={mockOnRemoveToast} />
    )

    // Fast-forward past the toast duration
    act(() => {
      jest.advanceTimersByTime(1100) // Just over 1 second
    })

    // The onOpenChange callback should be triggered, calling onClose
    expect(mockOnRemoveToast).toHaveBeenCalledWith('test-id-1')
  })

  test('uses default duration when not specified', () => {
    const mockOnRemoveToast = jest.fn()
    const mockToast: Toast = {
      id: 'test-id-1',
      title: 'Test Toast',
      variant: 'success',
      // No duration specified, should use default 3000ms
    }

    render(
      <ToastContainer toasts={[mockToast]} onRemoveToast={mockOnRemoveToast} />
    )

    // Fast-forward but not enough (shouldn't close)
    act(() => {
      jest.advanceTimersByTime(2000) // 2 seconds
    })
    expect(mockOnRemoveToast).not.toHaveBeenCalled()

    // Fast-forward past the default duration
    act(() => {
      jest.advanceTimersByTime(1500) // 3.5 seconds total
    })

    // Should be called now
    expect(mockOnRemoveToast).toHaveBeenCalledWith('test-id-1')
  })
})
