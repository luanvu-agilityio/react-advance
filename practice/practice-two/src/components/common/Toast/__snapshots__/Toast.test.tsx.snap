// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Component - Rendering matches snapshot for error toast 1`] = `
<div>
  <div
    aria-label="Notifications (F8)"
    role="region"
    style=""
    tabindex="-1"
  >
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
    <ol
      class="toast-viewport"
      tabindex="-1"
    >
      <li
        aria-atomic="true"
        aria-live="off"
        class="toast-root toast-error"
        data-radix-collection-item=""
        data-state="open"
        data-swipe-direction="right"
        role="status"
        style="user-select: none;"
        tabindex="0"
      >
        <div>
          <div
            class="toast-title"
            style="color: rgb(208, 0, 0);"
          >
            Error Toast
          </div>
          <div
            class="toast-description"
          >
            This is an error message
          </div>
        </div>
        <button
          class="toast-close"
          data-radix-toast-announce-exclude=""
          type="button"
        >
          ×
        </button>
      </li>
    </ol>
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Toast Component - Rendering matches snapshot for info toast 1`] = `
<div>
  <div
    aria-label="Notifications (F8)"
    role="region"
    style=""
    tabindex="-1"
  >
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
    <ol
      class="toast-viewport"
      tabindex="-1"
    >
      <li
        aria-atomic="true"
        aria-live="off"
        class="toast-root toast-info"
        data-radix-collection-item=""
        data-state="open"
        data-swipe-direction="right"
        role="status"
        style="user-select: none;"
        tabindex="0"
      >
        <div>
          <div
            class="toast-title"
            style="color: rgb(30, 136, 229);"
          >
            Info Toast
          </div>
        </div>
        <button
          class="toast-close"
          data-radix-toast-announce-exclude=""
          type="button"
        >
          ×
        </button>
      </li>
    </ol>
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Toast Component - Rendering matches snapshot for success toast 1`] = `
<div>
  <div
    aria-label="Notifications (F8)"
    role="region"
    style=""
    tabindex="-1"
  >
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
    <ol
      class="toast-viewport"
      tabindex="-1"
    >
      <li
        aria-atomic="true"
        aria-live="off"
        class="toast-root toast-success"
        data-radix-collection-item=""
        data-state="open"
        data-swipe-direction="right"
        role="status"
        style="user-select: none;"
        tabindex="0"
      >
        <div>
          <div
            class="toast-title"
            style="color: rgb(43, 147, 72);"
          >
            Success Toast
          </div>
          <div
            class="toast-description"
          >
            This is a success message
          </div>
        </div>
        <button
          class="toast-close"
          data-radix-toast-announce-exclude=""
          type="button"
        >
          ×
        </button>
      </li>
    </ol>
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Toast Component - Rendering matches snapshot for warning toast 1`] = `
<div>
  <div
    aria-label="Notifications (F8)"
    role="region"
    style=""
    tabindex="-1"
  >
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
    <ol
      class="toast-viewport"
      tabindex="-1"
    >
      <li
        aria-atomic="true"
        aria-live="off"
        class="toast-root toast-warning"
        data-radix-collection-item=""
        data-state="open"
        data-swipe-direction="right"
        role="status"
        style="user-select: none;"
        tabindex="0"
      >
        <div>
          <div
            class="toast-title"
            style="color: rgb(232, 93, 4);"
          >
            Warning Toast
          </div>
        </div>
        <button
          class="toast-close"
          data-radix-toast-announce-exclude=""
          type="button"
        >
          ×
        </button>
      </li>
    </ol>
    <span
      aria-hidden="true"
      style="position: fixed; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
      tabindex="0"
    />
  </div>
</div>
`;
